win32ss/drivers/font/ftfd/rosglue.c

```diff
@@ -36,52 +36,63 @@ DbgPrint(IN PCCH Format, IN ...)
  * buffer (need to copy the old contents to the new buffer). So, allocate
  * extra space for a size_t, store the allocated size in there and return
  * the address just past it as the allocated buffer.
+ * On win64 we need to align the allocation to 16 bytes, otherwise 8 bytes.
  */
+typedef struct _MALLOC_HEADER
+{
+    SIZE_T Size;
+    SIZE_T Alignment;
+} MALLOC_HEADER, * PMALLOC_HEADER;
 
 void *
 malloc(size_t Size)
 {
-    void *Object;
+    PMALLOC_HEADER Header;
 
-    Object = EngAllocMem(0, sizeof(size_t) + Size, TAG_FREETYPE);
-    if (Object != NULL)
+    Header = EngAllocMem(0, sizeof(MALLOC_HEADER) + Size, TAG_FREETYPE);
+    if (Header == NULL)
     {
-        *((size_t *)Object) = Size;
-        Object = (void *)((size_t *)Object + 1);
+        return NULL;
     }
 
-    return Object;
+    Header->Size = Size;
+    Header->Alignment = -1;
+    return (Header + 1);
 }
 
 void *
 realloc(void *Object, size_t Size)
 {
-    void *NewObject;
+    PMALLOC_HEADER OldHeader, NewHeader;
     size_t CopySize;
 
-    NewObject = EngAllocMem(0, sizeof(size_t) + Size, TAG_FREETYPE);
-    if (NewObject != NULL)
+    OldHeader = (PMALLOC_HEADER)Object - 1;
```

Comment:
`Object` can be `NULL`. See ANSI C.

Reply:
Specifying `NULL` pointer in `realloc` call works like `malloc`.
Specifying size zero in `realloc` call works like `free`.

Reply:
In realloc function, at first, if found NULL, then call malloc and return.
At second, if found size zero, then call free and return.
That will be ANSI C compliment.

Reply:
this is not ansi C, this is a shim for freetype.
Either way it might make sense to take into account that a realloc can be called with a `NULL Object`.

Reply:
if freetype happens to invoke realloc with these "ansi C" assumptions, then it may be necessary.
But it's clear that the old shim didn't seem to care, meaning freetype (at least the old versions) didn't invoke realloc that way.

Reply:
```suggestion
    if (!Size)
    {
        free(Object);
        return NULL;
    }
    if (!Object)
        return malloc(size);
    OldHeader = (PMALLOC_HEADER)Object - 1;
```

Reply:
I addressed the `Object == NULL` case. I didn't address `Size == 0`, because that is implementation defined behavior.